executor/aggfuncs/func_max_min.go

```diff
@@ -14,49 +14,174 @@
 package aggfuncs
 
 import (
+	"sort"
+
 	"github.com/pingcap/tidb/sessionctx"
 	"github.com/pingcap/tidb/types"
 	"github.com/pingcap/tidb/types/json"
 	"github.com/pingcap/tidb/util/chunk"
 	"github.com/pingcap/tidb/util/stringutil"
 )
 
+type maxMinQueue struct {
+	queue   []interface{}
+	varSet  map[interface{}]int64
+	isMax   bool
+	dirty   bool
+	cmpFunc func(i, j interface{}) int
+}
+
+func newMaxMinQueue(isMax bool) *maxMinQueue {
+	return &maxMinQueue{
+		queue:  make([]interface{}, 0),
+		varSet: make(map[interface{}]int64),
+		isMax:  isMax,
+	}
+}
+
+func (m *maxMinQueue) Resort() {
+	if len(m.queue) == 0 {
+		return
+	}
+	sort.Slice(m.queue, func(i, j int) bool {
+		if m.isMax {
+			return m.cmpFunc(m.queue[i], m.queue[j]) > 0
+		}
+		return m.cmpFunc(m.queue[i], m.queue[j]) < 0
+	})
+}
+
+func (m *maxMinQueue) Top() (val interface{}, isEmpty bool) {
+	if len(m.queue) == 0 {
+		return nil, true
+	}
+	if m.dirty {
+		m.Resort()
+		m.dirty = false
+	}
+	return m.queue[0], false
+}
+
+func (m *maxMinQueue) del(val interface{}) {
+	for i, qVal := range m.queue {
+		if m.cmpFunc(qVal, val) == 0 {
+			if i < len(m.queue) {
+				m.queue = append(m.queue[:i], m.queue[i+1:]...)
+			} else {
+				m.queue = m.queue[:i]
+			}
+		}
+	}
+}
+
+func (m *maxMinQueue) Push(val interface{}) {
+	if v, ok := m.varSet[val]; ok {
+		m.varSet[val] = v + 1
+	} else {
+		m.varSet[val] = 1
+		m.queue = append(m.queue, val)
+		m.dirty = true
+	}
+}
+
+func (m *maxMinQueue) PushMyDecimal(val interface{}) error {
```

Comment:
How about invoking `myDecimal.ToHashKey()` outside(in `maxMin4Decimal`) so that we can remove `PushMyDecimal`.

Reply:
I had remove `PushMyDecimal `, and append `MyDecimal` directly.  The memory use of these two implements is same.